extensions/workmanager/src/main/java/com/google/android/exoplayer2/ext/workmanager/WorkManagerScheduler.java

```diff
@@ -0,0 +1,164 @@
+import com.google.android.exoplayer2.util.Log;
+import com.google.android.exoplayer2.util.Util;
+
+import androidx.annotation.NonNull;
+import androidx.work.Constraints;
+import androidx.work.Data;
+import androidx.work.ExistingWorkPolicy;
+import androidx.work.NetworkType;
+import androidx.work.OneTimeWorkRequest;
+import androidx.work.WorkManager;
+import androidx.work.Worker;
+import androidx.work.WorkerParameters;
+
+/***
+ * A {@link Scheduler} that uses {@link WorkManager}.
+ */
+public final class WorkManagerScheduler implements Scheduler {
+
+    private static final String TAG = "WorkManagerScheduler";
+    private static final String KEY_SERVICE_ACTION = "service_action";
+    private static final String KEY_SERVICE_PACKAGE = "service_package";
+    private static final String KEY_REQUIREMENTS = "requirements";
+
+    private final String workName;
+
+    /**
+     * @param workName A name for work scheduled by this instance. If the same name was used by a previous
+     *     instance, anything scheduled by the previous instance will be canceled by this instance if
+     *     {@link #schedule(Requirements, String, String)} or {@link #cancel()} are called.
+     */
+    public WorkManagerScheduler(String workName) {
+        this.workName = workName;
+    }
+
+    @Override
+    public boolean schedule(Requirements requirements, String servicePackage, String serviceAction) {
+        Constraints constraints = buildConstraints(requirements);
+        Data inputData = buildInputData(requirements, servicePackage, serviceAction);
+        OneTimeWorkRequest workRequest = buildWorkRequest(constraints, inputData);
+        logd("Scheduling work: " + workName);
+        WorkManager.getInstance().enqueueUniqueWork(workName, ExistingWorkPolicy.REPLACE, workRequest);
+        return true;
+    }
+
+    @Override
+    public boolean cancel() {
+        logd("Canceling work: " + workName);
+        WorkManager.getInstance().cancelUniqueWork(workName);
+        return true;
+    }
+
+    private static Constraints buildConstraints(Requirements requirements) {
+        Constraints.Builder builder = new Constraints.Builder();
+
+        switch (requirements.getRequiredNetworkType()) {
+            case Requirements.NETWORK_TYPE_NONE:
+                builder.setRequiredNetworkType(NetworkType.NOT_REQUIRED);
+                break;
+            case Requirements.NETWORK_TYPE_ANY:
+                builder.setRequiredNetworkType(NetworkType.CONNECTED);
+                break;
+            case Requirements.NETWORK_TYPE_UNMETERED:
+                builder.setRequiredNetworkType(NetworkType.UNMETERED);
+                break;
+            default:
+                throw new UnsupportedOperationException();
+        }
+
+        if (requirements.isChargingRequired()) {
+            builder.setRequiresCharging(true);
+        }
+
+        if (requirements.isIdleRequired() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```

Comment:
Why version check is necessary?

Reply:
thats how the Builder API is defined
`        @RequiresApi(23)
        public @NonNull Builder setRequiresDeviceIdle(boolean requiresDeviceIdle) {
            this.mRequiresDeviceIdle = requiresDeviceIdle;
            return this;
        }`

Reply:
Ahh, true, I checked wrong method.
Please replace it with:
Util.SDK_INT >= 23